<launch>
  <node pkg="nodelet" type="nodelet" name="pcl_manager" args="manager" output="screen" />

  <!-- Run a passthrough filter to clean NaNs -->
  <node pkg="nodelet" type="nodelet" name="passthrough" args="load pcl/PassThrough pcl_manager" output="screen">
    <remap from="/passthrough/input"   to="/narrow_stereo_textured/points2" />
    <rosparam>
      filter_field_name: z
      filter_limit_min: 0.01
      filter_limit_max: 1.5
    </rosparam>
  </node>

  <!-- Estimate point normals -->
  <node pkg="nodelet" type="nodelet" name="normal_estimation" args="load pcl/NormalEstimation pcl_manager" output="screen">
    <remap from="/normal_estimation/input"   to="/passthrough/output" />
    <rosparam>
      # -[ Mandatory parameters
      # Set either 'k_search' or 'radius_search'
      k_search: 10
      radius_search: 0
      # Set the spatial locator. Possible values are: 0 (ANN), 1 (FLANN), 2 (organized)
      spatial_locator: 0
    </rosparam>
  </node>

  <!-- MUX the data onto one topic -->
  <node pkg="nodelet" type="nodelet" name="data_mux_1" args="load pcl/NodeletMUX pcl_manager" output="screen">
    <rosparam>
      input_topics: [/passthrough/output, /normal_estimation/output]
    </rosparam>
  </node>

  <!-- Concatenate fields -->
  <node pkg="nodelet" type="nodelet" name="concatenate_data" args="load pcl/PointCloudConcatenateFieldsSynchronizer pcl_manager" output="screen">
    <remap from="concatenate_data/input" to="/data_mux_1/output" />
    <rosparam>
      # Number of messages to expect on the input topic
      input_messages: 2
    </rosparam>
  </node>
  
</launch>

